### PR TITLE
LSP design proposal

### PR DESCRIPTION
**Pull request by 0101**
This is a design document for F# LSP. It's not exhaustive and definitely not final, but should be good to kick start discussions about the topic and base some initial work/investigation off of. 

Please check it out and comment/suggest changes and point out missing parts. 

To view rendered markdown go to Files changed, `...` -> View file

### CODE REVIEW

File: docs/lsp.md
Comment (baronfel) [inline docs/lsp.md:162]: IIRC FSAC does this automatic cancellation (at least of typechecking) and so this would be a take-back from users. I believe this matches 2+3 on your list here.

File: docs/lsp.md
Comment (baronfel) [inline docs/lsp.md:200]: StreamJsonRpc does keep track of in-flight requests for cancellation already. I think it's ok to lean on that library - it's pretty flexible and FSAC already relies on it.

File: docs/lsp.md
Comment (baronfel) [inline docs/lsp.md:246]: FSAC has background checking so this would be a take-back.

File: docs/lsp.md
Comment (baronfel) [inline docs/lsp.md:255]: I love this call-out :)

File: docs/lsp.md
Comment (baronfel) [inline docs/lsp.md:269]: One thing that the C# Dev Kit team has found is that there's not a lot of protocol-level support for performing the 'same kind' of fix at a whole-document or whole-project level, so if you want that you may have some custom work.

**Review by baronfel — COMMENTED** [review]
Left some very early thoughts. We should probably have a chat between the F# team and the FSAC maintainers about how things will work together.

File: docs/lsp.md
Comment (0101) [reply docs/lsp.md:162]: In the end the invalidated type check (expensive operation) needs to be canceled. This is more about whose responsibility it will be and which protocol for that will give best performance. Will dig deeper into it when we're implementing it.

File: docs/lsp.md
Comment (0101) [reply docs/lsp.md:246]: This is again more about who will control the background checking. It could be the client going through all files when it's idle and calling pull diagnostics (like it works now in VS). Or it could be the server and then we might use push diagnostics for it. 

The pull version seems more flexible to start with but open to discussion.

File: docs/lsp.md
Comment (majocha) [inline docs/lsp.md:206]: maybe https://github.com/VsixCommunity/Community.VisualStudio.Toolkit

File: docs/lsp.md
Comment (majocha) [inline docs/lsp.md:208]: It's very bare-bones. We would probably hit some missing functionality fast. Still, it can be used from a classic in-proc vsix and might provide some value.

Comment (vzarytovskii) [reply docs/lsp.md:208]: Last thing we want is to stay in proc. We don't need much of functionality. We pretty much only need an ability to talk to cps, Roslyn and lsp client

**Comment by nojaf** [comment]
Would this move FCS to something higher than `netstandard2.0`?

**Comment by auduchinok** [comment]
> Would this move FCS to something higher than netstandard2.0?

We still have to target `net472`/`netstandard2.0` in the Rider plugin for now, unfortunately. 😞

**Comment by vzarytovskii** [comment]
> Would this move FCS to something higher than `netstandard2.0`?

Not at this point. Once we are fully out of process, we can think about it.

**Comment by vzarytovskii** [comment]
> Would this move FCS to something higher than `netstandard2.0`?

Library on top of lsp will be net9 likely.

File: docs/lsp.md
Comment (majocha) [reply docs/lsp.md:208]: I'm not sure we can go purely oop with the extension. Things that come to mind:
* We provide functionality mostly by exporting F# flavored Roslyn interfaces. Can we do MEF export from VS.Extensibility?
* We need to provide some UI for user settings.
* We need to provide ClassificationDefinitions, i.e. syntax coloring.

On the bright side, none of these should hinder taking the checker oop, which is what matters.

Comment (vzarytovskii) [reply docs/lsp.md:208]: > I'm not sure we can go purely oop with the extension. Things that come to mind:
> 
> * We provide functionality mostly by exporting F# flavored Roslyn interfaces. Can we do MEF export from VS.Extensibility?
> 

If we do LSP we won't need many of those. For some cross-language things we will likely do delegating lsp or something similar. 

> * We need to provide some UI for user settings.
> 

That is supported without xaml, via unified settings APIs.

> * We need to provide ClassificationDefinitions, i.e. syntax coloring.
> 

This is possible for sure.

**Comment by vzarytovskii** [comment]
> Would this move FCS to something higher than `netstandard2.0`?

To add, given that long-term we want to:
* Host oop in VS.
* Ditch full framework compiler (as executable, not target of course), leave only coreclr one.

Once done, we can probably move to latest (-ish) target for FCS, so we can utilise things like Spans, and vectorized collection operations, etc.

File: docs/lsp.md
Comment (baronfel) [inline docs/lsp.md:39]: Can this point be clarified a bit? Right now it's unclear if the deliverable is a binary (i.e. that would supplant FSAC) or a library (that FSAC could consume to off-load parts of the protocol but still extend for things that we need).

Comment (vzarytovskii) [inline docs/lsp.md:39]: I don't really like this wording. I would prefer something like:
One implementation for LSP library, which can be used by all editors, including Visual Studio, which will cover the following:

LSP protocol and extensions handling
Workspace state handling (not the projects handling though, this is external)
Code fixes and refactorings
Delegating LSP if needed (to/from Roslyn)
Analyzers SDK plug-in (?)

Whereas the "server" (hosting it) part is orthogonal, and can either be handled by community projects (Ionide/FSAC) or by our simple LSP host (which we will be using in VS).

File: docs/lsp.md
Comment (vzarytovskii) [reply docs/lsp.md:206]: It is very unlikely we'll be able to take such big 3rd party dependency. Especially one which is VS-related, since we will be blocked each time VS updates extensibility layer (e.f. updates to dev18 or similar).

File: docs/lsp.md
Comment (vzarytovskii) [inline docs/lsp.md:57]:
```suggestion
VsCode(VS Code) --> Ionide(Ionide/FSAC)
```

File: docs/lsp.md
Comment (vzarytovskii) [inline docs/lsp.md:59]:
```suggestion
Vim(Vim) --> Ionide(Ionide/FSAC)
```

File: docs/lsp.md
Comment (vzarytovskii) [reply docs/lsp.md:39]: As I mentioned below, I think we can (and will do both), but we don't want to force ionide to use our server, but we would very much like to reuse majority of editor tooling (intelligence, diagnostics, inlay hints, quick fixes, refactoring, custom analyser messages) everywhere, so we don't duplicate efforts in supporting new features.

But I'll leave it to @0101 to clarify.

File: docs/lsp.md
Comment (vzarytovskii) [reply docs/lsp.md:246]: I do believe we should make it configurable - either bg worker to push diagnostics, or on-demand pull diagnostics on per-document basis.

File: docs/lsp.md
Comment (vzarytovskii) [inline docs/lsp.md:427]: I think priority of this one should be bumped.

File: docs/lsp.md
Comment (abonie) [reply docs/lsp.md:208]: > We pretty much only need an ability to talk to cps, Roslyn and **lsp client**

LSP server rather, right? Or am I misunderstanding?

Comment (vzarytovskii) [reply docs/lsp.md:208]: > > We pretty much only need an ability to talk to cps, Roslyn and **lsp client**
> 
> LSP server rather, right? Or am I misunderstanding?

Yeah. Well, thinking about it now, it's both. We need to make sure we can spawn LSP server (either as an API or external process, or both) as well as have an access to LSP client and register it for the F# and the server we spawn.

**Comment by nojaf** [comment]
What are the initial thoughts on the release process for these LSP deliverables?

**Comment by vzarytovskii** [comment]
> What are the initial thoughts on the release process for these LSP deliverables?
> 
> What would the cycle look like?

That is to be decided

File: docs/lsp.md
Comment (0101) [reply docs/lsp.md:246]: Yeah, we can do it like that eventually. But we can have a MVP with just pull diagnostics. I'll update it.

File: docs/lsp.md
Comment (0101) [reply docs/lsp.md:39]: Yeah I think we should provide both, I'll try to clarify it in the document. 

I'm envisioning that the library will contain a fully functional F# LSP server implementation that you can just connect to some in/out streams and it will work. And it's easy to also ship a binary that will just wrap that in an exe so it can be used directly.

But we should probably enable some extensibility so that you can add more endpoints or override existing ones to suit the needs of a particular editor. 

In an ideal world that shouldn't be needed much, as most functionality should be reusable by everyone and should just go directly into the library.

File: docs/lsp.md
Comment (0101) [reply docs/lsp.md:427]: Yeah I guess it will be helpful for development/testing also.

File: docs/lsp.md
Comment (0101) [reply docs/lsp.md]: I updated it. I think we need to say that this is for LSP _server_, because _client_ part won't really be included (or only for VS). But you can use it as a library.

Comment (0101) [reply docs/lsp.md]: I added a section on extensibility in the doc.

File: docs/lsp.md
Comment (vzarytovskii) [inline docs/lsp.md:367]: @0101
Based on yesterday's meeting with Cyrus, David and Todd, shall we add a columt which would mark whether certain requests supported by default VS client?

File: docs/lsp.md
Comment (vzarytovskii) [inline docs/lsp.md:241]: I guess based on discussion with Rolsyn folks, we'll likely go with CLaSP as base + maybe Ionide for types and additional helpers? We also should, for now, just ask EngSys to upload Ionide package to `dotnet-public`, so we can iterate faster.

File: docs/lsp.md
Comment (vzarytovskii) [inline docs/lsp.md:286]: Discussed it with Roslyn team as well, we will make sure it works in both VS and VSCode as soon as we have working LSP server in both.

File: docs/lsp.md
Comment (0101) [reply docs/lsp.md:367]: Yeah, I'll try to find out which ones they are. 

But I'm suspecting we won't be able to function just with whatever VS does by itself and we will have to add a custom client.

File: docs/lsp.md
Comment (0101) [reply docs/lsp.md:241]: It probably makes sense to try to prototype with it, see how awkward the API is from F#. Roslyn should also have the types somewhere which we could use, might be good enough for what we need them for.

Comment (vzarytovskii) [reply docs/lsp.md:241]: I mean the whole API is just registering handlers and types for messages, and that's it. It feels pretty slim and natural in F#. Roslyn types are not public I believe.

File: docs/lsp.md
Comment (vzarytovskii) [reply docs/lsp.md:367]: I think we should be good with most (if not all) **core** ones, since there are working LSP servers in VS right now, but might need to "mock" the client with MEF or ask LSP team to implement missing functionality.

File: docs/lsp.md
Comment (nojaf) [inline docs/lsp.md:229]: I'm quite sceptical about this approach. Perfecting the LSP transition for every F# project is an immense challenge. It's improbable that the team will perfect it on the first try, and waiting for the subsequent SDK for corrections won't motivate anyone to participate in this development. A nightly feed is essential here.

I earnestly urge the team to shape the LSP epic in a way that enables community members to actively participate and rectify issues specific to their codebases. While it's fine for the team to lead and make key decisions, they should genuinely strive to incorporate early contributors from the community into their process. Motivate individuals to embark on this journey with you. Simplify the testing and debugging processes for personal projects. This approach could address many unique issues through collective effort.

Regular updates, seamlessly integrated by vscode as an LSP client, could significantly accelerate progress. With ample documentation, some developers will find solutions for their specific needs.

**Review by abonie — APPROVED** [review]

**Review by abonie — APPROVED** [review]

**Review by T-Gro — APPROVED** [review]

**Pull request merged**
